### DISCUSSION TITLE
Can't start with shared docker containers on Macintosh

### DISCUSSION THREAD

**Discussion by deg** [verdaccio/verdaccio#2596] [Q&A] [answered]
Running Verdaccion on a Mac (Big Sur 11.6)

It starts fine if I don't try to mount volumes.  But, if I do, I get the ` fatal--- cannot open config file /verdaccio/conf/config.yaml: Error: CONFIG: it does not look like a valid config file`.

This is not an SELinux problem per see, since Macintosh does not have SELinux. But, maybe it is something similar?

Per the docs, I tried adding `:z`, but that did not help:

```
$ V_PATH=/Users/deg/verdaccio; docker run -it --rm --name verdaccio \ 
  -p 4873:4873 \
  -v $V_PATH/conf:/verdaccio/conf:z \
  -v $V_PATH/storage:/verdaccio/storage \
  -v $V_PATH/plugins:/verdaccio/plugins \
  verdaccio/verdaccio
 fatal--- cannot open config file /verdaccio/conf/config.yaml: Error: CONFIG: it does not look like a valid config file
```

 Do I need to set some Mac-specific permission?  Any other ideas?
 
 Thanks!

**Comment by juanpicado** [comment]
@deg look up the permissions folder, https://verdaccio.org/docs/docker#running-verdaccio-using-docker

The current message is not quite descriptive on the v5 of the real issue, this has been improved already on the next major but not yet shipped.

**Comment by deg** [comment]
That is the doc page that I was using.  But, I was not able to get it to work.
Maybe I'm misreading.  The only suggestions I saw on that page were to adjust SELinux (not relevant on a Mac) or add the :z flag.  Neither worked for me.

Am I misreading or misunderstanding something?

**Comment by deg** [comment]
Oh, now I see what you meant, I think.
I've now changed the permissions, but it still fails:

```
$ V_PATH=/Users/deg/verdaccio; docker run -it --rm --name verdaccio \
  -p 4873:4873 \
  -v $V_PATH/conf:/verdaccio/conf \
  -v $V_PATH/storage:/verdaccio/storage \
  -v $V_PATH/plugins:/verdaccio/plugins \
  verdaccio/verdaccio
 fatal--- cannot open config file /verdaccio/conf/config.yaml: Error: CONFIG: it does not look like a valid config file
 
$ ls -l verdaccio
 total 0
drwxr-xr-x  2 10001  65533  64 Oct 27 13:41 conf
drwxr-xr-x  2 10001  65533  64 Oct 27 13:41 plugins
drwxr-xr-x  2 10001  65533  64 Oct 27 13:41 storage

$ pwd
/Users/deg
```

**Reply by juanpicado** [reply]
🤔 I also have Big Sur and I just run this works fine, of course Docker requires permissions for mounting the folder, I'm using Docker Desktop.

```
~/projects/docker-test took 2s 
➜ V_PATH=/Users/jpicado/projects/docker-test; docker run -it --rm --name verdaccio3 \
  -p 4873:4873 \
  -v $V_PATH/conf:/verdaccio/conf \
  -v $V_PATH/storage:/verdaccio/storage \
  -v $V_PATH/plugins:/verdaccio/plugins \
  verdaccio/verdaccio
 warn --- config file  - /verdaccio/conf/config.yaml
 warn --- Plugin successfully loaded: verdaccio-htpasswd
 warn --- Plugin successfully loaded: verdaccio-audit
 warn --- http address - http://0.0.0.0:4873/ - verdaccio/5.2.0
 http --- 172.17.0.1 requested 'GET /'
 http --- 200, user: null(172.17.0.1), req: 'GET /', bytes: 0/1023
 http --- 172.17.0.1 requested 'GET /-/verdaccio/packages'
 http --- 200, user: null(172.17.0.1), req: 'GET /-/verdaccio/packages', bytes: 0/3
^C%                                                                                                                                                                                                         
~/projects/docker-test took 51s 
➜ ls -lsa
total 16
 0 drwxr-xr-x   6 jpicado  staff   192 Oct 27 12:47 .
 0 drwxr-xr-x  10 jpicado  staff   320 Oct 27 12:46 ..
16 -rw-r--r--@  1 jpicado  staff  6148 Oct 27 12:47 .DS_Store
 0 drwxr-xr-x   3 jpicado  staff    96 Oct 27 12:47 conf
 0 drwxr-xr-x   2 jpicado  staff    64 Oct 27 12:45 plugins
 0 drwxr-xr-x   3 jpicado  staff    96 Oct 27 12:53 storage

~/projects/docker-test 
➜ 
```

**Reply by deg** [reply]
Thanks.  That sure looks the same as what I did.
I'm far from expert at Docker on Mac (or even Mac in general) so let me dive into your comment "of course Docker requires permission..."

Do I need to set anything beyond these settings:
![image](https://user-images.githubusercontent.com/90383/139055128-a00002a7-ce38-44c6-a7f2-9d808e971aa2.png)

**Reply by juanpicado** [reply]
No as far I know 🙈 , I mean usually I don't have issues on Mac butI mostly work in Linux and 99% of the time is permissions, cold you do something crazy? `chmod -777 xxx` for all folders are being shared, just to be sure is permissions, it might be the `yaml` file you have in the `conf/` is not well formatted.

**Reply by deg** [reply]
`chmod 777` was the first thing I tried, and it didn't work.
But.... you just gave me a massive clue "it might be the yaml file is not well formatted".

Was I supposed to have created that file manually?  I thought it would be auto-created when the Docker container started.  (that certainly seemed to be the case when I started Docker without mounting volumes).

**Reply by juanpicado** [reply]
Yes, if you use volumes for `/conf` you need to create the file by your own :)

**Comment by deg** [comment]
Ok, thanks!!
What should be in the default file?
(or should I just copy the directory from the standalone docker container)

**Reply by juanpicado** [reply]
I'm not sure what you meant, but you have to copy https://github.com/verdaccio/verdaccio/blob/5.x/conf/docker.yaml into your new `conf/` folder named `config.yaml` otherwise you will get the same error.

**Comment by deg** [comment]
Thanks. Getting closer now.
I copied the file, and was able to start the Docker container.
I then tried adding a user, and got this in the log:
```
error--- the user deg could not being added. Error: EACCES: permission denied, open '/verdaccio/storage/htpasswd'
 error--- unexpected error: EACCES: permission denied, open '/verdaccio/storage/htpasswd'
Error: EACCES: permission denied, open '/verdaccio/storage/htpasswd'
 http --- 500, user: null(172.17.0.1), req: 'PUT /-/user/org.couchdb.user:deg', error: internal server error
```

Here are the file protections:

```
$ pwd
/Users/deg/verdaccio
 
$ ls -l
total 0
drwxr-xr-x  3 10001  65533  96 Oct 27 15:25 conf
drwxr-xr-x  2 10001  65533  64 Oct 27 13:41 plugins
drwxr-xr-x  2 10001  65533  64 Oct 27 13:41 storage
```

**Reply by juanpicado** [reply]
If you noticed, on my example I didn't have to use user container permissions, I'm not sure if is docker that allows that mapping, but this might be your issue now, maybe rollaback to the normal Mac user permissions for those folders and files.

**Comment by deg** [comment, marked as answer]
Success!
![image](https://user-images.githubusercontent.com/90383/139067053-d301f668-dada-4383-87e7-7bde9253e0ac.png)

Thank you, for this answer, for your amazingly fast responses, and for writing Verdaccio.  I look forward to using it now.  (Hope I don't have too many more questions!)

Suggestion for your docs: mention in the Docker page that you need to hand-create the yaml file if you use volumes.

**Reply by juanpicado** [reply]
It looks a an opportunity :) to became a contributor!! would you mind ? 

https://github.com/verdaccio/verdaccio/blob/master/website/docs/docker.md

(I'd suggest do it locally and ensure `pnpm format` command pass, prettier is quite picky with markdown.

**Comment by deg** [comment]
Sure.  Not sure I'll have time today, but happy to write up something.

**Comment by deg** [comment]
The PR is ready, but I don't have permission to push a branch to the repo. Do you want to add me, or should I just send you the text here?

**Reply by juanpicado** [reply]
Usually is easier as a fork, but since is ready  I've send you an invitation as collaborator :) welcome.

**Reply by deg** [reply]
I'm not seeing the invite yet, but no worries.  I PR'd from a fork now.
